alibi_detect/cd/base_online.py

```diff
@@ -290,7 +387,79 @@ def _configure_ref(self):
     def _update_state(self, x_t: np.ndarray):
         pass
 
+    def _set_state_dir(self, dirpath: Union[str, os.PathLike]):
+        """
+        Set the directory path to store state in, and create an empty directory if it doesn't already exist.
+
+        Parameters
+        ----------
+        dirpath
+            The directory to save state file inside.
+        """
+        self.state_dir = Path(dirpath)
+        self.state_dir.mkdir(parents=True, exist_ok=True)
+
+    def save_state(self, filepath: Union[str, os.PathLike]):
+        """
+        Save a detector's state to disk in order to generate a checkpoint.
+
+        Parameters
+        ----------
+        filepath
+            The directory to save state to.
+        """
+        self._set_state_dir(filepath)
+        self._save_state()
+        logger.info('Saved state for t={} to {}'.format(self.t, self.state_dir))
+
+    def load_state(self, filepath: Union[str, os.PathLike]):
+        """
+        Load the detector's state from disk, in order to restart from a checkpoint previously generated with
+        `save_state`.
+
+        Parameters
+        ----------
+        filepath
+            The directory to load state from.
+        """
+        self._set_state_dir(filepath)
+        self._load_state()
+        logger.info('State loaded for t={} from {}'.format(self.t, self.state_dir))
+
+    def _save_state(self):
+        """
+        Private method to save a detector's state to disk.
+
+        TODO - Method slightly verbose as designed to facilitate saving of "offline" state in follow-up PR.
+        """
+        filename = 'state'
+        keys = self.online_state_keys
+        save_state_dict(self, keys, self.state_dir.joinpath(filename + '.npz'))
+
+    def _load_state(self, offline: bool = False):
+        """
+        Private method to load a detector's state from disk.
+
+        TODO - Method slightly verbose as designed to facilitate loading of "offline" state in follow-up PR.
+        """
+        filename = 'state'
+        load_state_dict(self, self.state_dir.joinpath(filename + '.npz'), raise_error=True)
+
```

Comment:
Seems like a lot of duplicated code that's exactly the same as for `BaseMultiDriftOnline` which suggests we may want to refactor using functions instead of methods or a mixin class? Or perhaps the class hierarchy needs to be updated.

Reply:
Fair point. Having a  `BaseDriftOnline` class for generic methods, or a mix-in both sound much nicer than this. I'll have a rethink 👍🏻

Reply:
To reduce duplication, 5daf1b1 adds a `BaseDriftOnline` class. @jklaise @mauicv could I get your thoughts on the design of `BaseDriftOnline` please? I've gone with a parent class rather than mix-in since it seems strange to define a mix-in in `alibi_detect/base.py` when it is only to be used in two classes (`BaseMultiDriftOnline` and `BaseUniDriftOnline`). I also decided to put it in `alibi_detect/cd/base_online.py` rather than `alibi_detect/base.py` since at the moment the concept of "online" detectors is specific to drift (this may change if we decide stateful outlier detectors are in fact "online").

Reply:
LGTM however noting that there's quite a few abstract methods, some of which (not all?) are implemented in the `Multi/Uni` abstract methods, which come with their own set of abstract methods... Worried that this may become a bit tricky to keep track of. As a minimum, would group all abstract methods to come after each other and add docstrings on expected implementation and also, where valid, which of the `Multi/Uni` classes implement these methods (+ type hints as always).

Reply:
Tidied and moved to `StateMixin`. See https://github.com/SeldonIO/alibi-detect/pull/604#discussion_r1072154613.